src/BenchmarkDotNet/Configs/ConfigExtensions.cs

```diff
@@ -18,38 +18,90 @@ namespace BenchmarkDotNet.Configs
         /// </summary>
-        [PublicAPI] public static IConfig KeepBenchmarkFiles(this IConfig config, bool value = true) => config.With(m => m.Options = m.Options.Set(value, ConfigOptions.KeepBenchmarkFiles));
+        [Obsolete("This property will soon be removed, please start using WithOptionsIf instead.")]
+        [PublicAPI] public static IConfig KeepBenchmarkFiles(this IConfig config, bool value = true) => WithOptionsIf(config, value, ConfigOptions.KeepBenchmarkFiles);
         /// <summary>
         /// determines if the exported result files should not be overwritten (be default they are overwritten)
         /// </summary>
-        [PublicAPI] public static IConfig DontOverwriteResults(this IConfig config, bool value = true) => config.With(m => m.Options = m.Options.Set(value, ConfigOptions.DontOverwriteResults));
+        [Obsolete("This property will soon be removed, please start using WithOptionsIf instead.")]
+        [PublicAPI] public static IConfig DontOverwriteResults(this IConfig config, bool value = true) => WithOptionsIf(config, value, ConfigOptions.DontOverwriteResults);
         /// <summary>
         /// determines if benchmarking should be stopped after the first error (by default it's not)
         /// </summary>
-        [PublicAPI] public static IConfig StopOnFirstError(this IConfig config, bool value = true) => config.With(m => m.Options = m.Options.Set(value, ConfigOptions.StopOnFirstError));
-        [PublicAPI] public static IConfig WithArtifactsPath(this IConfig config, string artifactsPath) => config.With(m => m.ArtifactsPath = artifactsPath);
+        [Obsolete("This property will soon be removed, please start using WithOptionsIf instead.")]
+        [PublicAPI] public static IConfig StopOnFirstError(this IConfig config, bool value = true) => WithOptionsIf(config, value, ConfigOptions.StopOnFirstError);
+
         /// <summary>
         /// sets given options for the config
         /// </summary>
-        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => config.With(m => m.Options = options);
-        [PublicAPI] public static IConfig With(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => config.With(c => c.Add(rules));
+        [Obsolete("This property will soon be removed, please start using WithOptions instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => WithOptions(config, options);
+        [PublicAPI] public static ManualConfig WithOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithOptions(options));
+        [PublicAPI] public static ManualConfig WithoutOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithoutOptions(options));
+        [PublicAPI] public static ManualConfig WithOptionsIf(this IConfig config, bool value, ConfigOptions options) => config.With(m => m.WithOptionsIf(value, options));
```

Comment:
`WithOptionsIf` sounds pretty confusing to me.

Reply:
@AndreyAkinshin I've changed this name to `WithSwitchingOptions`. What do you think? 
I don't like them both, but I don't have a better idea.